src/test/unit/engine/InputTest.cpp

```diff
@@ -165,6 +169,548 @@ TEST(InputTest, Links)
+        true // isDefaultOutput
+        ));
+
+    return ns;
+  }
+
+  /**
+   * Inputs/Outputs are made available in initialize()
+   * It is always called after the constructor (or load from serialized state)
+   */
+  void initialize() override
+  {
+    nodeCount_ = getDimensions().getCount();
+    out_ = getOutput("out");
+    feedForwardIn_ = getInput("feedForwardIn");
+    lateralIn_ = getInput("lateralIn");
+  }
+
+  // Compute outputs from inputs and internal state
+  void compute() override
+  {
+    NTA_DEBUG << "> Computing: " << getName() << " <";
+
+    const Array & outputArray = out_->getData();
+    NTA_CHECK(outputArray.getCount() == 3);
+    NTA_CHECK(outputArray.getType() == NTA_BasicType_UInt64);
+    UInt64 *baseOutputBuffer = (UInt64*)outputArray.getBuffer();
+
+    std::vector<UInt64> ffInput;
+    feedForwardIn_->getInputForNode(0, ffInput);
+    NTA_CHECK(ffInput.size() > 1);
+
+    NTA_DEBUG << getName() << ".compute: ffInput size=" << ffInput.size()
+              << "; inputValue=" << ffInput[0];
+
+    std::vector<UInt64> latInput;
+    lateralIn_->getInputForNode(0, latInput);
+    NTA_CHECK(latInput.size() > 1);
+
+    NTA_DEBUG << getName() << ".compute: latInput size=" << latInput.size()
+              << "; inputValue=" << latInput[0];
+
+    baseOutputBuffer[0] = ffInput[0] + latInput[0];
+    baseOutputBuffer[1] = ffInput[0];
+    baseOutputBuffer[2] = latInput[0];
+
+    NTA_DEBUG << getName() << ".compute: out=" << baseOutputBuffer[0];
+  }
+
+private:
+  L2TestRegion();
+
+  /* ----- cached info from region ----- */
+  size_t nodeCount_;
+
+  // Input/output buffers for the whole region
+  const Input *feedForwardIn_;
+  const Input *lateralIn_;
+  const Output *out_;
+
+};
+
+
+class L4TestRegion: public TestRegionBase
+{
+public:
+  /*
+   * This region's output is computed as: k + feedbackIn
+   */
+  L4TestRegion(const ValueMap& params, Region *region) :
+    TestRegionBase(params, region),
+    k_(params.getScalarT<UInt64>("k"))
+  {
+  }
+
+  L4TestRegion(BundleIO& bundle, Region* region) :
+    TestRegionBase(bundle, region),
+    k_(0)
+  {
+  }
+
+  L4TestRegion(capnp::AnyPointer::Reader& proto, Region* region) :
+    TestRegionBase(proto, region),
+    k_(0)
+  {
+  }
+
+  virtual ~L4TestRegion()
+  {
+  }
+
+  std::string getNodeType()
+  {
+    return "L4TestRegion";
+  }
+
+  // Used by RegionImplFactory to create and cache
+  // a nodespec. Ownership is transferred to the caller.
+  static Spec* createSpec()
+  {
+    auto ns = new Spec;
+    /* ---- parameters ------ */
+    ns->parameters.add(
+      "k",
+      ParameterSpec(
+        "Constant k valud for output computation", // description
+        NTA_BasicType_UInt64,
+        1,                         // elementCount
+        "",                        // constraints
+        "",                        // defaultValue
+        ParameterSpec::ReadWriteAccess));
+
+    /* ----- inputs ------- */
+    ns->inputs.add(
```

Comment:
@ywcui1990, do you believe that the feedforward inputs from R1/R2 to R3/R4 don't adequately test the feedforward links?